commit dc717d47ad8c3eb2f5bd9148cf3dc9a9b4caa3e3
Author: David Clear <dac2@pensando.io>
Date:   Tue Mar 17 12:56:50 2020 -0700

    arch/arm64: Kconfig option to disable outer-cache-allocate.
    
    The Level 3 memory cache in the Pensando Capri SoC is used
    by the hardware data path cache objects and table data.  Forwarding
    performance can be increased by keeping the ARM cores out of L3.
    
    This commit adds the ARM64_MMU_DISABLE_OUTER_CACHE_ALLOCATE Kconfig
    option which changes how reads and writes are signaled on the
    core-external AXI interface.  Normal reads and writes that miss
    the L3 cache will be serviced from main memory but will not allocate
    a new cache line in L3; keeping L3 effectively reserved for the
    packet datapath.

diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index c4361df..732f493 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -1036,6 +1036,13 @@ config RANDOMIZE_MODULE_REGION_FULL
 	  a limited range that contains the [_stext, _etext] interval of the
 	  core kernel, so branch relocations are always in range.
 
+config ARM64_MMU_DISABLE_OUTER_CACHE_ALLOCATE
+	bool "Disable cache allocate in the outer domain"
+	depends on ARCH_PENSANDO
+        default n
+	help
+	  This option disables outer read/write cache-allocate"
+
 endmenu
 
 menu "Boot options"
diff --git a/arch/arm64/mm/proc.S b/arch/arm64/mm/proc.S
index 877d42f..c54a2e7 100644
--- a/arch/arm64/mm/proc.S
+++ b/arch/arm64/mm/proc.S
@@ -38,8 +38,13 @@
 
 #define TCR_SMP_FLAGS	TCR_SHARED
 
+#ifdef CONFIG_ARM64_MMU_DISABLE_OUTER_CACHE_ALLOCATE
+/* PTWs cacheable, inner WBWA, outer WBnWA */
+#define TCR_CACHE_FLAGS	TCR_IRGN_WBWA | TCR_ORGN_WBnWA
+#else
 /* PTWs cacheable, inner/outer WBWA */
 #define TCR_CACHE_FLAGS	TCR_IRGN_WBWA | TCR_ORGN_WBWA
+#endif
 
 #define MAIR(attr, mt)	((attr) << ((mt) * 8))
 
@@ -193,6 +198,27 @@ ENTRY(__cpu_setup)
 	isb					// Unmask debug exceptions now,
 	enable_dbg				// since this is per-cpu
 	reset_pmuserenr_el0 x0			// Disable PMU access from EL0
+
+#ifdef CONFIG_ARM64_MMU_DISABLE_OUTER_CACHE_ALLOCATE
+	/*
+	 * Memory region attributes for LPAE:
+	 *
+	 *   n = AttrIndx[2:0]
+	 *			n	MAIR
+	 *   DEVICE_nGnRnE	000	00000000
+	 *   DEVICE_nGnRE	001	00000100
+	 *   DEVICE_GRE		010	00001100
+	 *   NORMAL_NC		011	01000100
+	 *   NORMAL		100	11001111
+	 *   NORMAL_WT		101	10001011
+	 */
+	ldr	x5, =MAIR(0x00, MT_DEVICE_nGnRnE) | \
+		     MAIR(0x04, MT_DEVICE_nGnRE) | \
+		     MAIR(0x0c, MT_DEVICE_GRE) | \
+		     MAIR(0x44, MT_NORMAL_NC) | \
+		     MAIR(0xcf, MT_NORMAL) | \
+		     MAIR(0x8b, MT_NORMAL_WT)
+#else
 	/*
 	 * Memory region attributes for LPAE:
 	 *
@@ -211,6 +237,7 @@ ENTRY(__cpu_setup)
 		     MAIR(0x44, MT_NORMAL_NC) | \
 		     MAIR(0xff, MT_NORMAL) | \
 		     MAIR(0xbb, MT_NORMAL_WT)
+#endif
 	msr	mair_el1, x5
 	/*
 	 * Prepare SCTLR
